File: services/api/app/ai/embedding.py
"""Embedding helpers: tokenisation, deterministic embeddings, cosine similarity.

The deterministic embedding is a hashed bag-of-tokens projection. It is *not* a
learned semantic model, but it gives stable, meaningful overlap-based similarity
(texts sharing tokens are closer) which is perfect for offline development and
tests. In production the same interface is served by a real embedding model via
the configured provider.
"""

from __future__ import annotations

import hashlib
import itertools
import math
import re

from app.ai.providers.base import EMBEDDING_DIM

_ASCII_WORD = re.compile(r"[a-z0-9][a-z0-9+#.]*", re.IGNORECASE)
_CJK = re.compile(r"[\u4e00-\u9fff]")


def tokenize(text: str) -> list[str]:
    """Tokenise mixed Chinese/English text.

    Produces lowercase ASCII words plus single CJK characters and CJK bigrams,
    so Chinese phrases overlap meaningfully without a segmentation dependency.
    """
    if not text:
        return []
    lowered = text.lower()
    tokens: list[str] = [m.group(0) for m in _ASCII_WORD.finditer(lowered)]
    cjk_chars = _CJK.findall(text)
    tokens.extend(cjk_chars)
    tokens.extend(a + b for a, b in itertools.pairwise(cjk_chars))
    return tokens


def _bucket(token: str) -> tuple[int, float]:
    digest = hashlib.blake2b(token.encode("utf-8"), digest_size=8).digest()
    value = int.from_bytes(digest, "big")
    index = value % EMBEDDING_DIM
    sign = 1.0 if (value >> 8) & 1 else -1.0
    return index, sign


def deterministic_embedding(text: str, dim: int = EMBEDDING_DIM) -> list[float]:
    """Return a stable, L2-normalised embedding for ``text``."""
    vec = [0.0] * dim
    for token in tokenize(text):
        index, sign = _bucket(token)
        vec[index % dim] += sign
    norm = math.sqrt(sum(v * v for v in vec))
    if norm == 0.0:
        return vec
    return [v / norm for v in vec]


def cosine_similarity(a: list[float], b: list[float]) -> float:
    """Cosine similarity of two equal-length vectors, clamped to [0, 1]."""
    if not a or not b or len(a) != len(b):
        return 0.0
    dot = sum(x * y for x, y in zip(a, b, strict=False))
    na = math.sqrt(sum(x * x for x in a))
    nb = math.sqrt(sum(y * y for y in b))
    if na == 0.0 or nb == 0.0:
        return 0.0
    raw = dot / (na * nb)
    return max(0.0, min(1.0, raw))

File: apps/web/src/lib/ai/embedding.ts
/**
 * Deterministic, offline embeddings — a faithful TypeScript port of the backend
 * `app/ai/embedding.py`. The embedding is a hashed bag-of-tokens projection
 * (blake2b, 8-byte digest), so the same text yields the *same* vector as the
 * Python backend (verified byte-for-byte). No model or API key required.
 */
import { blake2b } from "@noble/hashes/blake2.js";

export const EMBEDDING_DIM = 256;

const ASCII_WORD = /[a-z0-9][a-z0-9+#.]*/gi;
const CJK = /[\u4e00-\u9fff]/g;
const encoder = new TextEncoder();

/**
 * Tokenise mixed Chinese/English text into lowercase ASCII words plus single
 * CJK characters and CJK bigrams, so Chinese phrases overlap meaningfully
 * without a segmentation dependency.
 */
export function tokenize(text: string): string[] {
  if (!text) return [];
  const lowered = text.toLowerCase();
  const tokens: string[] = [];
  for (const match of lowered.matchAll(ASCII_WORD)) tokens.push(match[0]);
  const cjkChars = text.match(CJK) ?? [];
  for (const ch of cjkChars) tokens.push(ch);
  for (let i = 0; i + 1 < cjkChars.length; i += 1) tokens.push(cjkChars[i] + cjkChars[i + 1]);
  return tokens;
}

function bucket(token: string): readonly [index: number, sign: number] {
  const digest = blake2b(encoder.encode(token), { dkLen: 8 });
  let value = 0n;
  for (const byte of digest) value = (value << 8n) | BigInt(byte);
  const index = Number(value % BigInt(EMBEDDING_DIM));
  const sign = ((value >> 8n) & 1n) === 1n ? 1 : -1;
  return [index, sign];
}

/** Return a stable, L2-normalised embedding for `text`. */
export function deterministicEmbedding(text: string, dim = EMBEDDING_DIM): number[] {
  const vec = new Array<number>(dim).fill(0);
  for (const token of tokenize(text)) {
    const [index, sign] = bucket(token);
    vec[index % dim] += sign;
  }
  let norm = 0;
  for (const v of vec) norm += v * v;
  norm = Math.sqrt(norm);
  if (norm === 0) return vec;
  return vec.map((v) => v / norm);
}

/** Cosine similarity of two equal-length vectors, clamped to [0, 1]. */
export function cosineSimilarity(a: number[], b: number[]): number {
  if (!a.length || !b.length || a.length !== b.length) return 0;
  let dot = 0;
  let na = 0;
  let nb = 0;
  for (let i = 0; i < a.length; i += 1) {
    dot += a[i] * b[i];
    na += a[i] * a[i];
    nb += b[i] * b[i];
  }
  if (na === 0 || nb === 0) return 0;
  const raw = dot / (Math.sqrt(na) * Math.sqrt(nb));
  return Math.max(0, Math.min(1, raw));
}
